internals/overlord/restart/restart.go

```diff
@@ -36,6 +36,7 @@ const (
 	RestartSystemHaltNow
 	// RestartSystemPoweroffNow will shutdown --poweroff the system asap
 	RestartSystemPoweroffNow
+	RestartServiceFailure
```

Comment:
Just sharing my mind here - maybe something we can boldly try to clean up a bit in the future ?

The ```RestartType``` name has evolved over time it seems to no longer work very well. The fact that we start the type with ```Restart``` prefix also does not help make this easy to follow. Seems like the meaning of many of these options are overloaded depending if you take the wider system effect into place (where systemd exists, and will restart Pebble on exit, or not). 

My view is this list tries to sometimes include 3 things:

Action, Scope and now also Reason (while having a prefix called Restart)

However, I cannot see an easy way to clean this up without mayor surgery (I read your spec considerations). In an ideal world, I would have referred to this group of actions and the package as ```shutdown``` (like the command), and perhaps a simpler list of options:

```go
type ShutdownType int
const (
    ShutdownHaltSystem ShutdownType = iota
    ShutdownRebootSystem
    ShutdownPoweroffSystem
    ShutdownExitDaemon                      // supports exit code
)

type Handler interface {
	HandleShutdown(t ShutdownType, immediate bool, exitcode int)
	RebootIsFine(st *state.State) error
	RebootIsMissing(st *state.State) error
}
```

Reply:
I don't disagree this is quite messy! However, I'm going to leave it for future work.

I've made a couple of tweaks to make the code a bit more like snapd's (which is even messier as it has more cases :-), for example renaming the field to `requestedRestart`.